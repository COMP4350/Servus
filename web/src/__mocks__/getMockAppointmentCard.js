function getMockAppointmentCard() {
    return {
        buyer: 'Risto',
        provider: 'Arveend',
        service_id: '1001',
        created_at: new Date(100),
        booked_time: new Date(500),
    }
}

export const mockAppointmentCard = getMockAppointmentCard();