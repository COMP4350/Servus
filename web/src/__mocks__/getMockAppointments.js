function getMockAppointments() {
    return [
        {
            buyer: 'Risto',
            provider: 'Arveend',
            service_id: '1001',
            created_at: new Date(100),
            booked_time: new Date(500),
        },
        {
            buyer: 'Arveend',
            provider: 'Risto',
            service_id: '1002',
            created_at: new Date(700),
            booked_time: new Date(1200),
        },
        {
            buyer: 'Caden',
            provider: 'Andy',
            service_id: '1003',
            created_at: new Date(1400),
            booked_time: new Date(1900),
        },
    ]
}

function getMockAppointmentsCookie() {
    return {
        username: 'testusername',
    }
}

export const mockAppointments = getMockAppointments();
export const mockAppointmentsCookie = getMockAppointmentsCookie();