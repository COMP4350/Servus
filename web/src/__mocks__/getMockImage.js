function getMockImage() {
    return {
        ownerUsername: 'riki',
        uploadDate: new Date(2012, 1, 1),
        imageName: 'example_image.jpeg',
        imageData: 'the/link/to/data',
        profilePicture: false,
    };
}

export const mockImage = getMockImage();
