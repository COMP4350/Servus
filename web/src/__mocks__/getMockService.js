function getMockService() {
    return {
        _id: '1001',
        name: "Risto's Painting",
        provider: 'Risto',
        description: 'Risto paints you',
        cost: 100,
        duration: '0100',
        availability: [{ weekday: 1, start_time: '0800', end_time: '1800' }],
        ratings: [{ username: 'arvind', rating: 5 }],
    };
}

export const mockService = getMockService();
