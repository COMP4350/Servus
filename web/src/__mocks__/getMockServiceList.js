function getMockServiceList() {
    return [
        {
            name: "Risto's Painting",
            provider: 'Risto',
            description: 'Risto paints you',
            cost: 100,
            duration: 60,
        },
        {
            name: "Reesto's Painting",
            provider: 'Reesto',
            description: 'Reesto paints yee',
            cost: 120,
            duration: 80,
        },
        {
            name: "Rasta's Painting",
            provider: 'Rasta',
            description: 'Rasta paints yaa',
            cost: 150,
            duration: 90,
        },
    ];
}

export const mockServiceList = getMockServiceList();
