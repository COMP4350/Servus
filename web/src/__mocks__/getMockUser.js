function getMockUser() {
    return {
        username: 'testuser',
        firstName: 'test',
        lastName: 'user',
        password: '123',
        bio: 'im a test user',
    };
}

export const mockUser = getMockUser();
