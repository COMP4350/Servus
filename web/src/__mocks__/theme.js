import { createMuiTheme } from '@material-ui/core/';

export default createMuiTheme({
    palette: {
        primary: {
            main: '#647AA3',
        },
        secondary: {
            main: '#EC5732',
        },
        default: {
            main: '#FFF2EB',
        },
    },
    background: {
        dark: '#151515',
        main: '#272727',
    },
    typographs: {
        fontFamily: ['Roboto'],
    },
});
