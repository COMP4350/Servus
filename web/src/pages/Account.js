import React, { useEffect, useState } from 'react';
import { makeStyles, createMuiTheme } from '@material-ui/core/styles';
import {
    Button,
    TextField,
    IconButton,
    ThemeProvider,
    Typography,
    Input,
} from '@material-ui/core';
import useForm from '../hooks/useForm';
import axios from 'axios';
import { useCookies } from 'react-cookie';
import ProfilePicture from '../components/ProfilePicture';
import { useHistory } from 'react-router';
import { ArrowBack } from '@material-ui/icons/';

const theme = createMuiTheme({
    palette: {
        type: 'dark',
        primary: {
            main: '#647AA3',
        },
        secondary: {
            main: '#EC5732',
        },
        default: {
            main: '#FFF2EB',
        },
    },
    background: {
        dark: '#151515',
        main: '#272727',
    },
    typography: {
        fontFamily: ['Roboto'],
    },
});

const useStyles = makeStyles(() => ({
    root: {
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'space-evenly',
        alignItems: 'center',
        margin: '2%',
    },
    title: {
        color: 'white',
    },
    textField: {
        width: '50%',
        marginBottom: '3%',
    },
    passwordTextField: {
        marginRight: '10px',
    },
    button: {
        textTransform: 'capitalize',
        marginRight: '10px',
    },
    buttonContainer: {
        display: 'flex',
        marginTop: '10px',
    },
    passwordContainer: {
        display: 'flex',
        marginBottom: '3%',
        alignItems: 'center',
    },
    accountBtn: {
        position: 'absolute',
        bottom: 0,
        left: 0,
        '&:hover': {},
    },
    arrowIcon: {
        height: '75px',
        width: '75px',
    },
    bio: {
        height: '100px',
        width: '50%',
        marginBottom: '20px',
    },
    pic: {
        color: 'white',
        width: 96,
        height: 96,
    },
}));

const Account = props => {
    const classes = useStyles();
    const history = useHistory();
    const [cookies, removeCookie] = useCookies(['username']);
    const [errors, setErrors] = useState({});
    const [passwordErrors, setPasswordErrors] = useState({});
    const [formValid, setFormValid] = useState();
    const [passwordValid, setPasswordValid] = useState();
    const [updatingPassword, setUpdatingPassword] = useState(false);
    const [form, onFormChange, setForm] = useForm({
        username: '',
        password: '',
        confirmPassword: '',
        firstName: '',
        lastName: '',
        bio: '',
    });
    const getUserInfo = () => {
        axios
            .get(`/user/${cookies.username}`)
            .then(res => {
                setForm({
                    username: res.data.result.username,
                    firstName: res.data.result.firstName,
                    lastName: res.data.result.lastName,
                    bio: res.data.result.bio,
                });
            })
            .catch(err => {
                alert('could not update password' + err);
            });
    };
    const validateInfo = () => {
        let errors = {};
        if (!form.firstName) {
            errors.username = 'first name is required';
        }
        if (!form.lastName) {
            errors.username = 'last name is required';
        }
        if (!form.username) {
            errors.username = 'username is required';
        }
        if (form.bio.length > 250) {
            errors.bio = 'bio must be under 250 characters';
        }
        setErrors(errors);
        setFormValid(Object.getOwnPropertyNames(errors).length == 0);
    };
    const logout = () => {
        removeCookie('username');
        props.setUsername('');
        history.push('/login');
    };
    const validatePassword = () => {
        let errors = {};
        if (!form.password) {
            errors.password = 'password is required';
        }
        if (!form.confirmPassword) {
            errors.confirmPassword = 'last name is required';
        } else if (form.password !== form.confirmPassword) {
            errors.confirmPassword = 'passwords must match';
        }
        setPasswordErrors(errors);
        setPasswordValid(
            Object.getOwnPropertyNames(errors).length == 0 ? true : false
        );
    };
    const updatePassword = async () => {
        if (passwordValid) {
            axios
                .put(`/user/${form.username}`, {
                    password: form.password,
                })
                .then(() => {
                    alert('Password updated successfully');
                    setUpdatingPassword(false);
                })
                .catch(err => {
                    alert('could not update password' + err);
                });
            setPasswordValid(false);
        }
    };
    const updateInfo = async () => {
        if (formValid) {
            axios
                .put(
                    `/user/${form.username}`,
                    {
                        password: form.password,
                        firstName: form.firstName,
                        lastName: form.lastName,
                        bio: form.bio,
                    },
                    {
                        withCredentials: true,
                    }
                )
                .then(() => {
                    alert('Info updated successfully');
                    props.setUsername(form.username);
                })
                .catch(err => {
                    alert(err);
                    setFormValid(false);
                });
            setFormValid(false);

            return;
        }
    };
    const toProfile = () => {
        history.push(`/profile/${cookies.username}`);
    };
    useEffect(() => {
        updateInfo();
    }, [formValid]);

    useEffect(() => {
        updatePassword();
    }, [passwordValid]);

    useEffect(() => {
        getUserInfo();
    }, []);
    const uploadImage = e => {
        let imageFormObj = new FormData();
        imageFormObj.append('imageName', 'multer-image-' + Date.now());
        imageFormObj.append('imageData', e.target.files[0]);
        imageFormObj.append('ownerUsername', cookies.username);
        imageFormObj.append('profilePicture', true);

        // stores a readable instance of the image being uploaded using multer
        axios
            .post(`/images/upload`, imageFormObj)
            .then(() => {})
            .catch(err => {
                alert('Error while uploading image' + err);
            });
    };

    return (
        <ThemeProvider theme={theme}>
            <div className={classes.root}>
                <Typography variant="h2" className={classes.title}>
                    Account Details
                </Typography>
                <ProfilePicture username={cookies.username} />
                <Input type="file" onChange={e => uploadImage(e, 'multer')} />
                <TextField
                    className={classes.textField}
                    id="outlined-basic"
                    label="First Name"
                    name="firstName"
                    value={form.firstName}
                    onChange={onFormChange}
                    error={errors.firstName}
                    helperText={errors.firstName}
                />
                <TextField
                    className={classes.textField}
                    id="outlined-basic"
                    label="Last Name"
                    name="lastName"
                    value={form.lastName}
                    onChange={onFormChange}
                    error={errors.lastName}
                    helperText={errors.lastName}
                />
                <TextField
                    className={classes.textField}
                    id="outlined-basic"
                    label="Username"
                    name="username"
                    value={form.username}
                    onChange={onFormChange}
                    error={errors.username}
                    helperText={errors.username}
                />
                <TextField
                    className={classes.bio}
                    id="outlined-basic"
                    label="Bio"
                    name="bio"
                    value={form.bio}
                    onChange={onFormChange}
                    multiline
                    rowsMax={4}
                    error={errors.bio}
                    helperText={errors.bio}
                />
                {updatingPassword ? (
                    <div className={classes.passwordContainer}>
                        <TextField
                            className={classes.passwordTextField}
                            id="outlined-password-input"
                            label="password"
                            type="Password"
                            autoComplete="current-password"
                            name="password"
                            value={form.password}
                            onChange={onFormChange}
                            error={passwordErrors.password}
                            helperText={passwordErrors.password}
                        />
                        <TextField
                            className={classes.passwordTextField}
                            id="outlined-password-input"
                            label="Confirm Password"
                            type="password"
                            name="confirmPassword"
                            value={form.confirmPassword}
                            onChange={onFormChange}
                            error={passwordErrors.confirmPassword}
                            helperText={passwordErrors.confirmPassword}
                        />
                        <Button
                            className={classes.button}
                            onClick={validatePassword}
                            variant="contained">
                            Update Password
                        </Button>
                    </div>
                ) : (
                    <Button
                        className={classes.button}
                        onClick={() => setUpdatingPassword(true)}
                        variant="contained">
                        Update Password
                    </Button>
                )}
                <div className={classes.buttonContainer}>
                    <Button
                        className={classes.button}
                        variant="contained"
                        onClick={validateInfo}>
                        Update Info
                    </Button>
                    <Button
                        className={classes.button}
                        variant="contained"
                        onClick={logout}>
                        Logout
                    </Button>
                </div>
                <IconButton
                    className={classes.accountBtn}
                    onClick={() => {
                        toProfile();
                    }}>
                    <ArrowBack className={classes.arrowIcon} />
                </IconButton>
            </div>
        </ThemeProvider>
    );
};

export default Account;
