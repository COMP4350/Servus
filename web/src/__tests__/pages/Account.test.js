import React from 'react';
import { act, render } from '@testing-library/react';
import Account from '../../pages/Account';
import { mockUser } from '../../__mocks__/getMockUser';
import mockAxios from 'axios';

jest.mock('react-cookie', () => {
    return {
        __esModule: true,
        useCookies: () => {
            return [{ username: 'testuser' }];
        },
    };
});

describe('Account', () => {
    it('renders correctly', async () => {
        mockAxios.get.mockImplementationOnce(() =>
            Promise.resolve({
                data: { result: mockUser },
            })
        );
        const promise = Promise.resolve();

        const { asFragment } = render(<Account />);

        expect(asFragment()).toMatchSnapshot();
        expect(mockAxios.get).toHaveBeenCalledWith('/user/testuser');

        await act(() => promise);
    });
});
