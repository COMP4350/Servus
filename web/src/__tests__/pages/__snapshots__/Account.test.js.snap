// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account renders correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1"
  >
    <h2
      class="MuiTypography-root makeStyles-title-2 MuiTypography-h2"
    >
      Account Details
    </h2>
    <div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root makeStyles-icon-13"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 5v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2H5c-1.11 0-2 .9-2 2zm12 4c0 1.66-1.34 3-3 3s-3-1.34-3-3 1.34-3 3-3 3 1.34 3 3zm-9 8c0-2 4-3.1 6-3.1s6 1.1 6 3.1v1H6v-1z"
        />
      </svg>
    </div>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline"
    >
      <input
        class="MuiInputBase-input MuiInput-input"
        type="file"
        value=""
      />
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root makeStyles-textField-3"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="outlined-basic"
        id="outlined-basic-label"
      >
        First Name
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          id="outlined-basic"
          name="firstName"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root makeStyles-textField-3"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="outlined-basic"
        id="outlined-basic-label"
      >
        Last Name
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          id="outlined-basic"
          name="lastName"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root makeStyles-textField-3"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="outlined-basic"
        id="outlined-basic-label"
      >
        Username
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          id="outlined-basic"
          name="username"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root makeStyles-bio-10"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink="false"
        for="outlined-basic"
        id="outlined-basic-label"
      >
        Bio
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
          id="outlined-basic"
          name="bio"
          rows="1"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
          tabindex="-1"
        />
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-5"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Update Password
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      class="makeStyles-buttonContainer-6"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-5"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Update Info
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-5"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Logout
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root makeStyles-accountBtn-8"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root makeStyles-arrowIcon-9"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
